src/qiboml/operations/expectation.py

```diff
@@ -128,11 +141,53 @@ def grad(upstream):
 
         return expval, grad
 
-    if differentiation_rule is not None:
-        return _expectation(params)
+    return _expectation(params)
 
-    elif nshots is None:
-        return _exact(observable, circuit, initial_state, exec_backend)
 
-    else:
-        return _with_shots(observable, circuit, initial_state, nshots, exec_backend)
+def _with_jax(
+    observable,
+    circuit,
+    initial_state,
+    nshots,
+    backend,
+    differentiation_rule,
+):
+    """
+    Compute expectation sample integrating the custom differentiation rule with
+    TensorFlow's automatic differentiation.
+    """
+    import jax  # pylint: disable=import-error
+
+    params = circuit.get_parameters()
+    exec_backend = construct_backend(backend)
+
+    @jax.custom_gradient
+    def _expectation(params):
+        params = jax.numpy.array(params)
+
+        def grad(params):
+            gradients = []
+            for p in range(len(params)):
+                gradients.append(
+                    differentiation_rule(
+                        circuit=circuit,
+                        hamiltonian=observable,
+                        parameter_index=p,
+                        initial_state=initial_state,
+                        nshots=nshots,
+                        backend=backend,
+                    )
+                )
```

Comment:
Batched execution wold be great here, for example, as you could perform psr on several different parameters parallely. Namely the batch would be 
```
((theta_0 + eps, theta_1, ..., theta_n),
(theta_0, theta_1 + eps, ..., theta_n),
...,
(theta_0, theta_1, ..., theta_n + eps))
```